File: RickMortyAndroidApp/main/src/main/kotlin/io/github/brunogabriel/rickmorty/main/characters/data/service/CharacterService.kt
package io.github.brunogabriel.rickmorty.main.characters.data.service

import io.github.brunogabriel.rickmorty.main.characters.data.models.CharacterResponse
import io.github.brunogabriel.rickmorty.shared.data.models.ListResponse
import retrofit2.http.GET
import retrofit2.http.Query

interface CharacterService {
    @GET("character")
    suspend fun getCharacters(
        @Query("page") page: Int
    ): ListResponse<CharacterResponse>
}

File: rick_and_morty_flutter/lib/character/data/service/character_service.dart
import 'package:dio/dio.dart';
import 'package:injectable/injectable.dart';
import 'package:rick_and_morty_flutter/character/data/response/character_response.dart';
import 'package:rick_and_morty_flutter/shared/network/api_response.dart';

@Injectable(as: ICharacterService)
class CharacterService implements ICharacterService {
  final Dio _client;

  const CharacterService(this._client);

  @override
  Future<ApiResponse<CharacterResponse>> getCharacters(int page) => _client
      .get("/character/?page=$page")
      .then((value) => CharacterApiResponse.fromJson(value.data));
}

abstract class ICharacterService {
  Future<ApiResponse<CharacterResponse>> getCharacters(int page);
}
